app/models/sftp/cli.rb

```diff
@@ -0,0 +1,266 @@
+    def execute_sftp_command(command)
+      stdin_data = "#{Array(command).join("\n")}\nquit\n"
+      output, status = Open3.capture2e(*build_command_parts, stdin_data: stdin_data)
+      return if status.success?
+
+      error_msg = output.lines.grep(/error|failed|denied|refused/i).join("\n")
+      error_msg = output if error_msg.empty?
+      raise StatusException, "SFTP command failed: #{command}\n#{error_msg}"
+    end
+
+    # Executes an sftp command and returns the output.
+    # Used for operations like 'ls' that need to parse the response.
+    def execute_sftp_with_output(command)
+      stdin_data = "#{Array(command).join("\n")}\nquit\n"
+      output, status = Open3.capture2e(*build_command_parts, stdin_data: stdin_data)
+      return output if status.success?
+
+      error_msg = output.lines.grep(/error|failed|denied|refused/i).join("\n")
+      error_msg = output if error_msg.empty?
+      raise StatusException, "SFTP command failed: #{command}\n#{error_msg}"
+    end
+
+    # Builds the command parts as an array for safe execution with Open3.
+    def build_command_parts
+      ensure_password_file if password_auth?
+
+      sftp_parts = ['sftp']
+      sftp_parts << '-P' << @options[:port].to_s if @options[:port]
+      if key_auth?
+        sftp_parts << '-i' << @options[:keys].first
+        sftp_parts << '-o' << 'PreferredAuthentications=publickey' if @options[:keys_only]
+      end
+      sftp_parts.concat(build_ssh_options_parts)
+      sftp_parts << "#{@username}@#{@host}"
+
+      if password_auth?
+        ['sshpass', '-f', @password_file.path] + sftp_parts
+      else
+        sftp_parts
+      end
+    end
+
+    # Builds SSH option flags as separate -o flag=value pairs for the sftp command.
+    def build_ssh_options_parts
+      opts = []
+      opts << '-o' << 'StrictHostKeyChecking=no' if @options[:skip_verify_host_key]
+      if @options[:keepalive]
+        interval = @options[:keepalive_interval] || 60
+        opts << '-o' << "ServerAliveInterval=#{interval}"
+        opts << '-o' << 'ServerAliveCountMax=3'
+      end
+      opts
+    end
+
+    # Returns true if password authentication should be used.
+    def password_auth?
+      @options[:password].present? || @options[:auth_methods]&.include?('password')
+    end
+
+    # Returns true if SSH key authentication should be used.
+    def key_auth?
+      @options[:keys].present? && @options[:keys].any?
+    end
+
+    # Creates a temporary file with restricted permissions (0600) containing the password.
+    # This file is used by sshpass to avoid exposing the password in process lists or command history.
+    def ensure_password_file
+      return if @password_file && File.exist?(@password_file.path)
+
+      @password_file = Tempfile.new('sftp_password')
+      @password_file.write(@options[:password] || '')
+      @password_file.close
+      File.chmod(0o600, @password_file.path)
+    end
```

Comment:
password written to Tempfile before permissions are set creates a brief window where the file may be readable

```suggestion
    def ensure_password_file
      return if @password_file && File.exist?(@password_file.path)

      @password_file = Tempfile.new('sftp_password', mode: 0o600)
      @password_file.write(@options[:password] || '')
      @password_file.close
    end
```

Reply:
Thanks! `Tempfile.new` doesn't take `mode` but I've moved the chmod above the write, hopefully that's sufficient!